[feature] Add DP Random Forest Classifier algorithm

mismayil opened the pull request:
I would like to add an implementation of a differentially private Random Forest Classifier algorithm based on [this paper](https://arxiv.org/abs/1606.03572).  As mentioned in [this issue](https://github.com/IBM/differential-privacy-library/issues/22), I read the [survey](https://dl.acm.org/doi/pdf/10.1145/3337064) on the differentially private decision forest algorithms and according to the authors, the paper used in this PR is one of the best-performing methods in terms of prediction accuracy and incurred privacy loss. In addition to the unit tests, I have tested the method privately on a real-world dataset at work as well. 

Since there was no documentation on contribution guidelines, I am not sure if my PR meets all the requirements. Happy to provide more detail on the implementation and update the PR. CC: @naoise-h

gkaissis commented:
This looks correct, why isn't this being merged?

naoise-h commented:
Hi, thanks for the reminder. We're hoping to be in a position to review this in the coming weeks, as time allows.

Best wishes, Naoise.

naoise-h reviewed (changes requested):
Hi Mete,

Thank you very much for your contribution. Apologies it has taken this long to get to.

Having taken a first look through the code, I have the following comments.

1. It would be great to refactor the code so that it fits better within the sklearn ecosystem. For example, all the classifiers should be instances of sklearn's `BaseEstimator` class at a minimum. This has many benefits, including be able to leverage existing sklearn functionality. From what I can see, the majority of the code you have can be maintained, but refactored a bit. It may not be possible to subclass the `RandomForestClassifier`, but it should be possible to subclass `ForestClassifier` at least.

2. Similarly, your `DecisionTreeClassifier` should be a subclass of `BaseDecisionTree` at least, if not a subclass of `DecisionTreeClassifier`. Again, this helps us leverage existing sklearn functionality, and ensures easier integration.

3. Finally, and this isn't an issue with your implementation, but rather the paper itself. I'm not sure if they have implemented smooth sensitivity as it should be. You typically can't use a smooth sensitivity bound as you would a global sensitivity bound. For example, while you can use Laplace noise with global sensitivity to achieve (pure) differential privacy, smooth sensitivity requires Cauchy noise to satisfy pure differential privacy. I'm not aware of a proof showing smooth sensitivity being used as a replacement for global sensitivity in the exponential mechanism. (In fact, it seems I can construct examples that break differential privacy when using a set-up of the exponential mechanism proposed in this paper)

4. The `get_feature_domains` function seems to be a non-private function whose outputs are used extensively in the training process. I would suggest allowing a user input specifying these domains (thereby making it privacy-preserving), or else throwing a `PrivacyLeakWarning` when/if it is read from the data. It is an open problem on how best to estimate these values in a privacy-preserving way (#20).


I would be keen to get your thoughts on these before we proceed further.

Thanks,
Naoise.

sam-fletcher commented:
Hello, I'm the author of the original paper. Thank you Mete for putting in the work to implement our algorithm in this fantastic library! 

Naoise has raised a good point in (3), that I'm in the process of addressing. I expect to have a solution for you by this Wednesday.

mismayil commented:
Hi Naoise,

Thank you for reviewing the PR! I am currently on vacation and will be addressing your comments soon when I am back. 

And Sam, it was my pleasure to work on this implementation and hope to get it merged together with your help! Let me know once you address the concern and I will incorporate it into my PR.

mismayil commented:
Hi @naoise-h 

I am back and I got some time to go through your review comments. I think those are good suggestions and I made new changes to address all of them except the 3rd one which I am still working on. With these new changes, `RandomForestClassifier` and `DecisionTreeClassifier` inherit from `ForestClassifier` and `BaseDecisionTree` respectively. Parent classes are initialized with overlapping attributes. Both classes also accept a new parameter called `feature_domains` and if not provided, a privacy leak warning is raised. Tests have been added to test new functionalities as well. While I am working on adjusting for Sam's correction, could you please review the changes and let me know if they address your comments well?

P.S. The new lgtm alert seems to be due to the fact that sklearn `ClassifierMixin` and `MultiOutputMixin` (subclassed by `BaseDecisionTree`) both have the method called `_more_tags`. I added inheritance from both similar to what sklearn does for `DecisionTreeClassifier`, but if you think we don't need `ClassifierMixin`, I can remove that. 

Thanks,
Mete

naoise-h commented:
This looks great Mete, thanks. I'll do my best to take a proper look tomorrow.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+        warn_unused_args(unused_args)
+
+    @property
+    def n_features(self):
+        return self._n_features
+
+    @property
+    def n_classes(self):
+        return self._n_classes
+
+    @property
+    def classes_(self):
+        return self._classes
+
+    @property
+    def cat_features(self):
+        return self._cat_features
+
+    @property
+    def max_depth(self):
+        return self._max_depth
+
+    @property
+    def estimators(self):
+        return self._estimators
+
+    @property
+    def feature_domains(self):
+        return self._feature_domains
```
These need not be defined, as they can be accessed directly from the class variable.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        X, y = np.array(X), np.array(y)
+        check_X_y(X, y)
+
+        self._n_features = X.shape[1]
```
Instead of `self._n_features`, use `self.n_features_in_`, as `n_features_` will raise an error in some versions of scikit-learn.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        self._accountant.check(self._epsilon, 0)
+
+        X, y = np.array(X), np.array(y)
+        check_X_y(X, y)
```
Since we're no requiring Scikit-learn 0.23, you can use `self._validate_data(X, y)` here instead.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+            return True
+        return False
+
+    def predict(self, X):
```
Let's try and get rid of this `predict()` method, since we can use the inherited method. This means we need a `predict_proba()` method in `DecisionTree()`. I'll investigate next week, but for our case this should mean returning a binary array with `1` at the index that we're predicting.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        return np.apply_along_axis(lambda x: Counter(list(x)).most_common(1)[0][0], arr=y, axis=1)
+
+
+class DecisionTreeClassifier(ClassifierMixin, BaseDecisionTree):
```
To solve the LGTM error, just define it's own `_more_tags()` method.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+            min_impurity_decrease=None,
+            min_impurity_split=None
+        )
+        self._feature_domains = feature_domains
+        self._cat_features = None
+        self._classes = None
+        self._cat_feature_threshold = cat_feature_threshold
+        self._max_depth = max_depth
+        self._epsilon = epsilon
+        self._root = None
+        self._fitted = False
```
Same as for the main class, parameters shouldn't have a leading underscore, fitted attributes shouldn't be defined here, and should have a trailing underscore.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+        return self
+
+    def check_is_fitted(self):
```
Once you've refactored the fitted attributes as described, you can use scikit-learn's `check_is_fitted` from `sklearn.utils.validation` instead. Just note, it either raises an NotFittedError or returns None, so you don't need to run a boolean check on it.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        if random_state is not None:
+            np.random.seed(random_state)
+
+    def set_cat_features(self, cat_features):
+        self._cat_features = cat_features
+
+    def set_classes(self, classes):
+        self._classes = classes
```
Can we specify these as parameters at initialisation instead?

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+        return self
+
+    def predict(self, X):
```
We can do away with this method by using `predict()` from `BaseDecisionTree`. We just need the underlying tree to have a `predict()` method.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        n_features = X.shape[1]
+        features = list(range(n_features))
+
+        self._root = self._build(features, self._feature_domains)
```
Can we call it `self.tree_` to fit with `BaseDecisionTree`?

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+        return np.array(y)
+
+    def check_is_fitted(self):
```
Let's use `check_is_fitted` from `sklearn.utils.validation` instead, which can be done one we refactor the fitted attribute names as described above.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+    """
+
+    def __init__(self, n_estimators=10, epsilon=1.0, cat_feature_threshold=10,
+                 n_jobs=1, verbose=0, accountant=None, max_depth=15, random_state=None, feature_domains=None, **unused_args):
```
Can reformat these lines to fit within the 120-char line length?

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+
+        if self._feature_domains is None:
+            warnings.warn(
+                "`feature_domains` parameter hasn't been specified, so falling back to determining domains from the data.\n"
```
This line is outside the 120-char line length limit too.

naoise-h commented on diffprivlib/models/__init__.py:
```diff
@@ -24,3 +24,4 @@
 from diffprivlib.models.logistic_regression import LogisticRegression
 from diffprivlib.models.pca import PCA
 from diffprivlib.models.standard_scaler import StandardScaler
+from diffprivlib.models.tree import RandomForestClassifier
```
Can we change the filename to `forest.py` to match the equivalent in scikit-learn?

naoise-h commented on tests/models/test_tree.py:
```diff
@@ -0,0 +1,173 @@
+                model.fit(X, y)
+
+
+class TestDecisionTreeClassifier(TestCase):
```
Can we split this into different files, one for each class that's being tested?

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+        if self._n_estimators > len(X):
+            raise ValueError('Number of estimators is more than the available samples')
+
+        subset_size = int(len(X) / self._n_estimators)
```
General comment here: If our dataset has 99 entries, and we have 10 estimators, this will result in 9 estimators with 9 entries, and the last estimator with 18 entries? Perhaps there's a better way to do this, but not crucial for a first draft to be merged.

sam-fletcher replied:
I think this line is a holdover from my original repo. The intent is for each tree to have a roughly-equal portion of the data. The subsets need to be disjoint (no overlap) in order for the privacy guarantees to hold.

naoise-h commented on diffprivlib/models/tree.py:
```diff
@@ -0,0 +1,602 @@
+            random_state=random_state,
+            max_leaf_nodes=None,
+            min_impurity_decrease=None,
+            min_impurity_split=None
```
It seems that `min_impurity_split` has been removed for scikit-learn v1.0, so you can remove it from here.

naoise-h replied:
After looking at this again, it's a required argument for prior versions, so this will need to be a try...except statement.

naoise-h reviewed (changes requested):
Thanks again Mete, I've gone through it in some detail, and have suggested changes so it fits better within the scikit-learn framework. It's mostly cosmetic changes that should be easy to do in any IDE.

Simple things like defining a `predict` method in the base decision tree will reduce the code we need to add, and adoping scikit-learn's attribute naming convention will make it a bit more streamlined too (even though some of the conventions aren't ideal re PEP 8). 

There's quite a lot here, so if you need some help, feel free to give me write access to your repo and let me know!

mismayil commented:
Thanks, @naoise-h  for your thorough review. I made some more changes to address most of your comments and marked them as 'Resolved' for easier visibility. Will take a look at others later and in the meantime, you can review if you are good with the resolutions.

naoise-h commented on diffprivlib/models/forest.py:
```diff
@@ -0,0 +1,598 @@
+                    if val not in self._class_counts:
+                        self._class_counts[val] = 0
+
+                if max([v for k, v in self._class_counts.items()]) < 1:
+                    self._noisy_label = np.random.choice([k for k, v in self._class_counts.items()])
+                else:
+                    all_counts = sorted([v for k, v in self._class_counts.items()], reverse=True)
+                    count_difference = all_counts[0] - all_counts[1]
+                    self._sensitivity = np.exp(-1 * count_difference * epsilon)
+                    self._noisy_label = self._calc_label_by_expo_mech(epsilon, self._sensitivity, self._class_counts)
```
To implement Sam's changes, you can change these lines to the following, and delete the `_calc_label_by_expo_mech` method.
```
utility = list(self._class_counts.values())
candidates = list(self._class_counts.keys())
mech = PermuteAndFlip(epsilon=epsilon, sensitivity=1, monotonic=True, utility=utility, candidates=candidates)
self.noisy_label_ = mech.randomise()
```

mismayil replied:
Thanks, Naoise. I will add that in.

sam-fletcher replied:
@naoise-h is there any risk of exponential floating point errors when class counts are high? Is it worth subtracting the smallest class count from each count, so that we baseline the utilities at zero? 

Not sure if its a problem in practice though.

naoise-h replied:
You're correct Sam, but this is handled within the mechanism itself, here:

https://github.com/IBM/differential-privacy-library/blob/main/diffprivlib/mechanisms/exponential.py#L235

mismayil commented:
Hi @naoise-h,

I think with these last changes, I have addressed all your comments, however, the build seems to be failing after adding your suggestion for Sam's fixes. Tests were successful on my machine though:) Could you please take a look as well?

naoise-h reviewed (changes requested):
It looks like you've addressed everything in the previous review, thanks Mete!

To help the tests pass, whenever you're doing an assert on a `predict()`, you can increase epsilon to 5. You need not worry about the codecov failing, this is a config error on our part.

There are some warnings in the tests that aren't being caught: [link](https://app.travis-ci.com/github/IBM/differential-privacy-library/jobs/538268055#L965). Either catch the warnings with `@pytest.mark.filterwarnings`, or refactor the test so a warning isn't triggered in the first place.

Last thing is to fix some linting issues. You can ignore W0201,W0221 and C0103 errors, but the others should be simple to fix: [link](https://app.travis-ci.com/github/IBM/differential-privacy-library/jobs/538268055#L333)

Once these changes are made, the PR will be ready to merge!

naoise-h reviewed (approved):
LGTM, thanks for your contribution Mete!

mismayil commented:
> LGTM, thanks for your contribution Mete!

My pleasure! Thanks for reviewing and accepting my contribution, Naoise!

sam-fletcher commented:
Thank you again @mismayil for taking the initiative and implementing my algorithm, and thank you @naoise-h for noticing the missing proof!
